utils/hare_cov/hare_coverage.py

```diff
@@ -0,0 +1,362 @@
+        """
+        Check for the existence of the file system node.
+        :param dir_path: file or directory path
+        :return: bool
+        """
+        return path.exists(dir_path)
+
+    def get_abs_hare_home_path(self) -> str:
+        """
+        Search for the absolute path from where the script can search for
+        the test and source directory for unit testcase and coverage run.
+        :return: absolute hare home path in string.
+        """
+        absolute_file_path = path.realpath(__file__)
+        pos = absolute_file_path.find(f"utils{sep}hare_cov{sep}"
+                                      "hare_coverage.py")
+        if pos != -1:
+            return absolute_file_path[:pos]
+        else:
+            logging.error('Not able to get relative hare home path.')
+            print('Not able to get relative hare home path.')
+            exit(1)
+
+
+class Strategy(ABC):
+    """Abstract base class for coverage tool."""
+
+    @abstractmethod
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        pass
+
+    @abstractmethod
+    def check_report(self) -> None:
+        """
+        Validate the report generated
+        :return: None
+        """
+        pass
+
+
+class PythonCoverage(Strategy):
+
+    # This is source to test directory map.It will be used run coverage.
+    # All the newly added source and test should be added here for coverage
+    # calculation.
+    source_test_map_hare: Dict[str, str] = {
+        "hax/hax": "hax/test/",
+        "provisioning/miniprov/hare_mp": "provisioning/miniprov/test/"
+    }
+
+    def __init__(self, cov_report_type: str = "xml",
+                 report_dir: str = "hare_coverage_report",
+                 st_map: Dict[str, str] = source_test_map_hare.copy()):
+        """
+        Sets coverage report type, directory where report will generated and
+        unit test directories source modules.
+        :param cov_report_type: coverage type xml or html
+        :param report_dir: directory where report will generated
+        :param st_map: source module to test module map
+        """
+        self.cov_report_type: str = cov_report_type
+        self.src_tst_map: Dict[str, str] = st_map
+        self.utils = Utility()
+        self.abs_path = self.utils.get_abs_hare_home_path()
+        self.report_dir = self.abs_path + report_dir
+        self.report = self.report_dir + sep
+        self.report += (
+            "html_report" if self.cov_report_type == "html"
+            else "coverage.xml")
+
+        logging.info("Running Python Coverage Tool.")

```

Comment:
In my case the logs don't look very informative:
```
2021-08-24 01:57:37,652 [INFO] Running Python Coverage Tool.
2021-08-24 01:57:37,652 [INFO] Cleaning up the report.
2021-08-24 01:57:37,652 [INFO] Running UT's.
2021-08-24 01:58:05,115 [INFO] Running Coverage.
2021-08-24 01:58:33,300 [INFO] Checking report.
2021-08-24 01:58:33,300 [INFO] Python Code Coverage Report generated successfully at '/mnt/projects/cortx-hare/hare_coverage_report/coverage.xml'.
2021-08-24 01:58:33,301 [INFO] Running C Coverage Tool.
```


> Running Python/C Coverage Tool.

Nice to know. But can I see a bit more details? Test names or folders or something?

> 2021-08-24 01:57:37,652 [INFO] Running UT's.

Why "UT"? I've been developing the software for more than 10 years. I can't say that UT is a well-known and widely used abbreviation. Let's replace it with "unit tests" to make the message clear.

BTW which unit tests are we running? Can we list any details in the log?

> 2021-08-24 01:58:33,300 [INFO] Checking report.

It is not clear what "checking" stands for and why it can be interesting to the reader. Feels like this is more like a DEBUG message, not INFO.


At the end it is not clear whether any tests were found, whether they failed or not and what is the verdict (is my code good enough in terms of test coverage?). Can we get any text report from `coverage` printed as a part of the log?

Reply:
Done